conclave-host/src/main/kotlin/com/r3/conclave/host/internal/StreamHostEnclaveInterface.kt

```diff
@@ -0,0 +1,218 @@
+package com.r3.conclave.host.internal
+
+import com.r3.conclave.common.internal.*
+import com.r3.conclave.mail.internal.readInt
+import com.r3.conclave.utilities.internal.getAllBytes
+import java.io.InputStream
+import java.io.OutputStream
+import java.nio.ByteBuffer
+import java.util.concurrent.ArrayBlockingQueue
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Semaphore
+
+/**
+ * This class is a streaming implementation of the [HostEnclaveInterface].
+ * It has three jobs:
+ *  - Serve as the endpoint for calls to make to the enclave, see [com.r3.conclave.common.internal.CallInterface]
+ *  - Route calls from the enclave to the appropriate host side call handler, see [com.r3.conclave.common.internal.CallInterface]
+ *  - Handle the low-level details of the messaging protocol (streamed ecalls and ocalls).
+ */
+class StreamHostEnclaveInterface(
+        private val outputStream: OutputStream,     // Messages going to the enclave
+        private val inputStream: InputStream,       // Messages arriving from the enclave
+) : HostEnclaveInterface() {
+    private var isRunning = true
+
+    /** Receive the number of concurrent calls from the enclave. */
+    private val maxConcurrentCalls = inputStream.readInt()
+    private val callGuardSemaphore = Semaphore(maxConcurrentCalls)
+
+    private val receiveLoop = object : Runnable {
```

Comment:
Why do we need to receive messages from the enclave on a background thread? Since the enclave never calls the host outside the context of an host->enclave call, we can simply block on the input stream waiting for a response from the enclave after the host has sent its message. If I understand the code correctly, that on its own would reduce a lot of the complexity in this class.

Reply:
Multiple threads need to be able to call the enclave simultaneously!

Reply:
But you can't send messages down the socket stream simultaneously, as evidenced by your locking of the streams. So we might as well simplify things and block on the receiving stream, waiting for the enclave's response. This is another reason why the socket implementation is probably temporary and we will need to look into shared memory for example at some point.

Reply:
As per our conversation, I've added a comment to the send methods.